### PR TITLE
[TE] added a couple of validations to check if detection is subscribed & valid; cleaned up other validations

### PR DESCRIPTION
**Pull request by akshayrai**
Validations added:
* Make sure subscription group subscribes to the detectionName while creating an alert
* Make sure all the subscribed detections are valid
* recipients should be configured under the EMAIL scheme params and not at the root level

Validations removed:
* type is not compulsory in subscription group
* type is not compulsory for sub-entity alerts.

Added unit tests for newly added validations

### CODE REVIEW

**Comment by xiaohui-sun** [comment]
Thanks for the clean up. I am thinking of refactoring the validation module to:
1. Can expose multiple error messages together.
2. Easier to add more validation rules.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java [hunk 1]
```diff
@@ -262,6 +263,23 @@ private Response processBadAuthorizationResponse(String type, String operation,
...
+    Preconditions.checkArgument(config.containsKey(PROP_DETECTION), "Detection pipeline yaml is missing");
+    Preconditions.checkArgument(config.containsKey(PROP_SUBSCRIPTION), "Subscription group yaml is missing.");
+
+    // check if subscription group has subscribed to the detection
+    Map<String, Object> detectionConfigMap = new HashMap<>(ConfigUtils.getMap(this.yaml.load(config.get(PROP_DETECTION))));
+    String detectionName = MapUtils.getString(detectionConfigMap, PROP_DETECTION_NAME);
+    Map<String, Object> subscriptionConfigMap = new HashMap<>(ConfigUtils.getMap(this.yaml.load(config.get(PROP_SUBSCRIPTION))));
+    List<String> detectionNames = ConfigUtils.getList(subscriptionConfigMap.get(PROP_DETECTION_NAMES));
+    Preconditions.checkArgument(detectionNames.contains(detectionName),
+        "You have not subscribed to the alert. Please copy-paste the detectionName under the "
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: There are some use cases (e.g, DHM) that they only want to create alert w/o subscription groups.

Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: There is already a separate api to create just the detection.

**Comment by akshayrai** [comment]
> 1. Can expose multiple error messages together.

That might be better but I am sure how much value it will bring. We can club together some independent validations and show them together but we still won't be able to run validation on the complete config.

> 2. Easier to add more validation rules.

The validations are already abstracted out and written in a separate validation class. What are you thinking about?

**Comment by xiaohui-sun** [comment]
> > 1. Can expose multiple error messages together.
> 
> That might be better but I am sure how much value it will bring. We can club together some independent validations and show them together but we still won't be able to run validation on the complete config.
We want to show all validation results to users. Currently any error will block the other validators.
> 
> > 1. Easier to add more validation rules.
> 
> The validations are already abstracted out and written in a separate validation class. What are you thinking about?
It is pretty good now. Some improvement we can do: 
1. Pass the error message more clearly. Some of the error message was wrapped in exceptions. Some of them are in precondition checks.
2. The validators are coupled and states are shared. You can't run the validators separately. 
3. Some validations are warnings and some are errors. We should differentiate them.

**Comment by akshayrai** [comment]
@xiaohui-sun these are great ideas. We can discuss them offline and take that up as a separate effort and not merge into this?

**Comment by akshayrai** [comment]
Validations added:
* Make sure subscription group subscribes to the detectionName while creating an alert
* Make sure all the subscribed detections are valid
* recipients should be configured under the EMAIL scheme params and not at the root level

Validations removed:
* type is not compulsory in subscription group
* type is not compulsory for sub-entity alerts.

Added unit tests for newly added validations

**Pull request merged**